zilliqa/src/consensus.rs

```diff
@@ -974,16 +983,18 @@ impl Consensus {
 
         let transactions = self.get_txns_to_execute();
 
-        let mut gas_used = EvmGas(0);
+        let mut gas_left = BLOCK_GAS_LIMIT;
         let applied_transactions: Vec<_> = transactions
             .into_iter()
             .filter_map(|tx| {
                 self.apply_transaction(tx.clone(), parent_header, inspector::noop())
                     .transpose()
                     .map(|r| {
-                        r.map(|result| {
-                            gas_used += result.gas_used();
-                            tx
+                        r.and_then(|result| {
+                            gas_left = gas_left
+                                .checked_sub(result.gas_used())
+                                .ok_or_else(|| anyhow!("too much gas used"))?;
```

Comment:
If this anyhow() fires, does anyone then propose a new block? (in what way does the next block to be proposed differ st the next potential leader won't fire this anyhow! again?)

Reply:
The intention is that this `anyhow!` never fires, thanks to the checks in `get_txns_to_execute`. If it does fire (because I am mistaken and there is a bug), this node won't propose a block and the other nodes will vote for a new view (after a timeout). Whether the same bug occurs again depends on the bug i guess. Though in theory once this node is the proposer again, the transactions which caused problems will no longer be in its mempool.